io/src/main/scala/fs2/io/tcp/Socket.scala

```diff
@@ -116,18 +117,18 @@ protected[tcp] object Socket {
     }
 
     def connect(ch: AsynchronousSocketChannel): F[AsynchronousSocketChannel] =
-      F.async { cb =>
+      F.async[AsynchronousSocketChannel] { cb =>
         ch.connect(
           to,
           null,
           new CompletionHandler[Void, Void] {
             def completed(result: Void, attachment: Void): Unit =
-              invokeCallback(cb(Right(ch)))
+              cb(Right(ch))
             def failed(rsn: Throwable, attachment: Void): Unit =
-              invokeCallback(cb(Left(rsn)))
+              cb(Left(rsn))
           }
         )
-      }
+      } <* cs.shift
```

Comment:
I think we could do 
```scala
def yieldBack = F.start(F.unit).flatMap(_.join)
```
and only require a `Concurrent` constraint, which imho better represents what we're doing here.

Reply:
Ah, and probably we want to yield back in a `bracket`? (although I appreciate this is a POC PR :) )

Reply:
@SystemFw Fixed now, let me know what you think